pkg/reconciler/source/crd/crd.go

```diff
@@ -0,0 +1,308 @@
+	}
+
+	return nil
+}
+
+func (r *Reconciler) finalize(ctx context.Context, crd *v1beta1.CustomResourceDefinition) error {
+	// The Source CRD is being deleted, we need to delete the dynamically created controller...
+	gvr, gvk, err := r.resolveGroupVersions(ctx, crd)
+	if err != nil {
+		logging.FromContext(ctx).Error("Error while resolving GVR", zap.String("CRD", crd.Name), zap.Error(err))
+		return err
+	}
+	r.lock.RLock()
+	rc, found := r.controllers[*gvr]
+	r.lock.RUnlock()
+	if found {
+		r.lock.Lock()
+		// Now that we grabbed the write lock, check that nobody deleted it already.
+		rc, found = r.controllers[*gvr]
+		if found {
+			logging.FromContext(ctx).Info("Stopping Source Duck Controller", zap.String("GVR", gvr.String()), zap.String("GVK", gvk.String()))
+			rc.cancel()
+			delete(r.controllers, *gvr)
+		}
+		r.lock.Unlock()
+	}
+	return nil
+}
+
+func (r *Reconciler) resolveGroupVersions(ctx context.Context, crd *v1beta1.CustomResourceDefinition) (*schema.GroupVersionResource, *schema.GroupVersionKind, error) {
+	var gvr *schema.GroupVersionResource
+	var gvk *schema.GroupVersionKind
+	for _, v := range crd.Spec.Versions {
+		if !v.Served {
+			continue
+		}
+		gvr = &schema.GroupVersionResource{
+			Group:    crd.Spec.Group,
+			Version:  v.Name,
+			Resource: crd.Spec.Names.Plural,
+		}
+
+		gvk = &schema.GroupVersionKind{
+			Group:   crd.Spec.Group,
+			Version: v.Name,
+			Kind:    crd.Spec.Names.Kind,
+		}
+
+	}
+	if gvr == nil || gvk == nil {
+		return nil, nil, fmt.Errorf("unable to find GVR or GVK for %s", crd.Name)
+	}
+	return gvr, gvk, nil
+}
+
+func (r *Reconciler) reconcileController(ctx context.Context, crd *v1beta1.CustomResourceDefinition, gvr *schema.GroupVersionResource, gvk *schema.GroupVersionKind) error {
+	r.lock.RLock()
+	rc, found := r.controllers[*gvr]
+	r.lock.RUnlock()
+	if found {
+		return nil
+	}
+
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	// Now that we grabbed the write lock, check that nobody has created the controller.
+	rc, found = r.controllers[*gvr]
+	if found {
+		return nil
+	}
+
+	// Source Duck controller constructor
+	sdc := duck.NewController(crd.Name, *gvr, *gvk)
+	// Source Duck controller context
+	sdctx, cancel := context.WithCancel(r.ogctx)
+	// Source Duck controller instantiation
+	sd := sdc(sdctx, r.ogcmw)
+
+	rc = runningController{
+		controller: sd,
+		cancel:     cancel,
+	}
+	r.controllers[*gvr] = rc
+
+	logging.FromContext(ctx).Info("Starting Source Duck Controller", zap.String("GVR", gvr.String()), zap.String("GVK", gvk.String()))
+	go func(c *controller.Impl) {
+		if err := c.Run(controller.DefaultThreadsPerController, sdctx.Done()); err != nil {
+			logging.FromContext(ctx).Error("Unable to start Source Duck Controller", zap.String("GVR", gvr.String()), zap.String("GVK", gvk.String()))
+		}
+	}(rc.controller)
+	return nil
+}
+
+func (r *Reconciler) addFinalizer(ctx context.Context, crd *v1beta1.CustomResourceDefinition) (*v1beta1.CustomResourceDefinition, error) {
```

Comment:
I really do not think we want this

Reply:
ok, will remove. If something gets removed, then we will still have that controller in memory, but I guess is not a big deal